makeflow/src/makeflow_hook.c

```diff
@@ -55,45 +72,93 @@ struct dag_file * makeflow_hook_add_output_file(struct dag *d, struct batch_task
 }
 
 
-int makeflow_hook_register(struct makeflow_hook *hook) {
+int makeflow_hook_register(struct makeflow_hook *hook, struct jx **args) {
 	assert(hook);
-	if (!makeflow_hooks) makeflow_hooks = list_create();
+	if (!makeflow_hooks)     makeflow_hooks     = list_create();
+	if (!makeflow_hook_args) makeflow_hook_args = list_create();
+	if (!makeflow_hook_self) makeflow_hook_self = list_create();
 
 	/* Add hook by default, if it doesn't exists in list of hooks. */
 	int rc = MAKEFLOW_HOOK_SUCCESS;
 	struct makeflow_hook *h = NULL;
-	debug(D_MAKEFLOW_HOOK, "Hook %s:trying to registered",hook->module_name?hook->module_name:"");
+	struct jx *h_args = NULL;
+	debug(D_MAKEFLOW_HOOK, "Hook %s:trying to register",hook->module_name?hook->module_name:"");
 
 	if(hook->register_hook){
-		rc = hook->register_hook(hook, makeflow_hooks);
-	} else {
-		list_first_item(makeflow_hooks);
-		while((h = list_next_item(makeflow_hooks))){
-			if(!strcmp(h->module_name, hook->module_name)){
+		rc = hook->register_hook(hook, makeflow_hooks, args);
+	}
+
+	if(!(hook->register_hook) || rc == MAKEFLOW_HOOK_SKIP) {
+		struct list_cursor *cur = list_cursor_create(makeflow_hooks);
+		struct list_cursor *acur = list_cursor_create(makeflow_hook_args);
+		// This should be a reverse traversal so that the most recent
+		// hook of the same name is used.
+		// Now it will always fully traverse the list to get the last
+		// instance, which is logically the same as the first reverse
+		// instance.
+		for (list_seek(cur, 0), list_seek(acur, 0); 
+			 list_get(cur, (void**)&h) && list_get(acur, (void**)&h_args); 
+			 list_next(cur), list_next(acur)){
+			if(h && !strcmp(h->module_name, hook->module_name)){
+				*args = h_args;
 				rc = MAKEFLOW_HOOK_SKIP;
-				break;
 			}
 		}
+		list_cursor_destroy(cur);
+		list_cursor_destroy(acur);
 	}
 
 	if(rc == MAKEFLOW_HOOK_SUCCESS){
-		h = xxmalloc(sizeof(*h));
-		memcpy(h, hook, sizeof(*h));
-		debug(D_MAKEFLOW_HOOK, "Hook %s:registered",h->module_name?h->module_name:"");
+		h_args = jx_object(NULL);
+		if(h_args){
+			*args = h_args;
+		}
+
+		list_push_tail(makeflow_hooks, hook);
+		list_push_tail(makeflow_hook_args, h_args);
+		list_push_tail(makeflow_hook_self, NULL);
 
-		list_push_tail(makeflow_hooks, h);
+		debug(D_MAKEFLOW_HOOK, "Hook %s:registered",hook->module_name?hook->module_name:"");
 	} else if(rc == MAKEFLOW_HOOK_FAILURE){
-		debug(D_MAKEFLOW_HOOK, "Hook %s:register failed",h->module_name?h->module_name:"");
-	} else if(rc == MAKEFLOW_HOOK_SKIP){
-		debug(D_MAKEFLOW_HOOK, "Hook %s:register skipped",h->module_name?h->module_name:"");
+		/* Args are NULL to prevent other hooks from modifying an
+		 * unintended arg list. */
+		args = NULL;
```

Comment:
Did you mean to dereference this (and delete the JX)? This will only clear the local copy of the pointer, not the value.